pkg/ebpf/uprobes/attacher.go

```diff
@@ -0,0 +1,901 @@
+	// ExcludeInternal excludes internal DataDog processes
+	ExcludeInternal
+	// ExcludeBuildkit excludes buildkitd processes
+	ExcludeBuildkit
+	// ExcludeContainerdTmp excludes containerd tmp mounts
+	ExcludeContainerdTmp
+)
+
+var (
+	// ErrSelfExcluded is returned when the PID is the same as the agent's PID.
+	ErrSelfExcluded = errors.New("self-excluded")
+	// ErrInternalDDogProcessRejected is returned when the PID is an internal datadog process.
+	ErrInternalDDogProcessRejected = errors.New("internal datadog process rejected")
+	// ErrNoMatchingRule is returned when no rule matches the shared library path.
+	ErrNoMatchingRule = errors.New("no matching rule")
+	// regex that defines internal DataDog processes
+	internalProcessRegex = regexp.MustCompile("datadog-agent/.*/((process|security|trace)-agent|system-probe|agent)")
+)
+
+// AttachTarget defines the target to which we should attach the probes, libraries or executables
+type AttachTarget uint8
+
+const (
+	// AttachToExecutable attaches to the main executable
+	AttachToExecutable AttachTarget = 1 << iota
+	// AttachToSharedLibraries attaches to shared libraries
+	AttachToSharedLibraries
+)
+
+// ProbeOptions is a structure that holds the options for a probe attachment. By default
+// these values will be inferred from the probe name, but the user can override them if needed.
+type ProbeOptions struct {
+	// IsManualReturn indicates that the probe is a manual return probe, which means that the inspector
+	// will find the return locations of the function and attach to them instead of using uretprobes.
+	IsManualReturn bool
+
+	// Symbol is the symbol name to attach the probe to. This is useful when the symbol name is not a valid
+	// C identifier (e.g. Go functions)
+	Symbol string
+}
+
+// AttachRule defines how to attach a certain set of probes. Uprobes can be attached
+// to shared libraries or executables, this structure tells the attacher which ones to
+// select and to which targets to do it.
+type AttachRule struct {
+	// LibraryNameRegex defines which libraries should be matched by this rule
+	LibraryNameRegex *regexp.Regexp
+	// ExecutableFilter is a function that receives the path of the executable and returns true if it should be matched
+	ExecutableFilter func(string, *ProcInfo) bool
+	// Targets defines the targets to which we should attach the probes, shared libraries and/or executables
+	Targets AttachTarget
+	// ProbesSelectors defines which probes should be attached and how should we validate
+	// the attachment (e.g., whether we need all probes active or just one of them, or in a best-effort basis)
+	ProbesSelector []manager.ProbesSelector
+	// ProbeOptionsOverride allows the user to override the options for a probe that are inferred from the name
+	// of the probe. This way the user can set options such as manual return detection or symbol names for probes
+	// whose names aren't valid C identifiers.
+	ProbeOptionsOverride map[string]ProbeOptions
+}
```

Comment:
Are all options can co-exists?

If not, how can we validate / ensure corrected of the rules?

Reply:
Yes, having rules that can attach to both shared libraries and executables is useful when targeting libraries that sometimes are compiled statically. In any case the `Validate` method below is called later by the attacher on creation to check that everything is valid.